### DISCUSSION TITLE
Convert Chinese name to Pinyin in citation key.

### DISCUSSION THREAD

**Discussion by earthmoon** [retorquere/zotero-better-bibtex#2360] [General]
Hello,
When the authors' names are Chinese characters, the Better Bibtex would not insert Chinese characters in the citation keys, whereas in the citation key setting they should contain the author's name. 
One of these solutions is in the Better Bibtex setting in Zotero, choose **Ideographs in citekeys** under the **Citation keys** tab. This way is good for the Japanese, but not good for the Chinese, as it will convert Chinese characters to Japanese ideographs. So, I hope Better Bibtex will develop a function for converting Chinese characters to Pinyin in the citation keys, as Chinese always understand Pinyin.
Thanks.

![Screenshot 2023-01-03 105545](https://user-images.githubusercontent.com/45897471/210294702-65d714b0-0b8d-44ca-a832-73a7c26fc209.jpg)

**Comment by retorquere** [comment]
To be sure I'd need a debug log, but I think `auth.fold(chinese).lower + year` will work.

**Reply by earthmoon** [reply]
Thanks for your help. The way of `auth.fold(chinese).lower + year` works well. However I can't reproduce the problem, as I turn off the function of **Ideographs in citekeys** Zotero will convert Chinese characters to Pinyin under the condition of setting `auth.lower + year` after I had setting `auth.fold(chinese).lower + year` once. So, only a totally new condition can reproduce this problem. I will do this in the future if I have a chance.

**Reply by earthmoon** [reply]
Before, Better Bibtex will leave Chinese characters empty.

**Reply by retorquere** [reply]
I'm terribly sorry, but I'm having difficulty determining what you are trying to say above.

**Reply by earthmoon** [reply]
> I'm terribly sorry, but I'm having difficulty determining what you are trying to say above.

Let me say them in detail:

1. The problem is very common as we all have seen this problem in China;
2. Zotero can't generate citation keys with Pinyin which contains Chinese by `auth.lower + year`.  But Zotero will generate citation keys as 20191 or 20192 and so on, which omit any Pinyin about Chinese characters. This is the problem I want to show;
3. As I saw this problem, I choose the function of  **Ideographs in citekeys**, which can convert Chinese to ideographs. It is better than the format of 20191 or 20192;
4. Then, I did as you taught me: `auth.fold(chinese).lower + year`, and I found the Chinese names can be converted to Pinyin correctly;
5. Then, I chose not to use **Ideographs in citekeys**, set `auth.lower + year`, and opened the debug mode in Zotero. I wanted to reproduce the problem and give you the debug ouput;
6. Then I found Zotero with Better Bibtex can convert Chinese names to Pinyin, in which situation I can't reproduce the problem;
7. But this problem is displayed on other computers. If I have a chance to see it, I will try to output the debug log and upload it here;

Those are what I wanted to say. I hope my poor English wouldn't confuse you again.
Thanks.

**Reply by retorquere** [reply]
`Ideographs in citekeys` is a header above two checkboxes, but it is not a setting itself. Those checkboxes activate `kuroshiro` and/or `jieba`. I think if you disable both, restart Zotero, and try to refresh the citekey, you will see the behavior you describe in point 2. If you also disable `Force citation key to plain text` you ought to get citekeys with chinese characters in them.

A debug log will give me a copy of the selected items so we know we're looking at the same data. You don't have to wait until you have an error condition.

**Reply by earthmoon** [reply]
> `Ideographs in citekeys` is a header above two checkboxes, but it is not a setting itself. Those checkboxes activate `kuroshiro` and/or `jieba`. I think if you disable both, restart Zotero, and try to refresh the citekey, you will see the behavior you describe in point 2. If you also disable `Force citation key to plain text` you ought to get citekeys with chinese characters in them.
> 
> A debug log will give me a copy of the selected items so we know we're looking at the same data. You don't have to wait until you have an error condition.

Hello,
I did as you suggested.

1. Set key format as: `auth.lower + year`
2. I disabled both check of `Ideographs in citekeys`;
3. I disabled `Force citation key to plain text`;
4. I refreshed the keys of Chinese papers, and it generated Chinese keys;
5. I enabled `Force citation key to plain text`;
6. I refreshed the keys of Chinese papers, and it generated Pinyin keys;

I can't reproduce the behavior in point2. 
This is debug information:

> D1988719960


> version => 6.0.19, platform => Win32, oscpu => Windows NT 10.0; WOW64, locale => en-US, appName => Zotero, appVersion => 6.0.19, extensions => Zotero LibreOffice Integration (6.0.3.SA.6.0.19, extension), Zotero Word for Windows Integration (6.0.2.SA.6.0.19, extension), Better BibTex for Zotero (6.7.43, extension), Zotero PDF Preview (0.0.2, extension), Sci-Hub Plugin for Zotero (1.4.4, extension)

**Reply by retorquere** [reply]
I don't have access to `D<number>` logs, please right-click an entry for which the problem occurs, select `Better BibTeX` -> `Send debug log`, and you'll get a log ID with `-refs-` in it.

**Reply by earthmoon** [reply]
Hello,
I have done it. My log ID is: 
[NRPWIZMR-refs-apse.zip](https://github.com/retorquere/zotero-better-bibtex/files/10337877/NRPWIZMR-refs-apse.zip)
NRPWIZMR-refs-apse

**Reply by retorquere** [reply]
My bad. Turning on `Force` will always transliterate, not remove, so I'm not sure how you got `20191` earlier. For `NRPWIZMR-refs-apse` I get `WeiXiao2016` when I turn `Force` on with `auth.lower + year`. The `lower` doesn't do anything because `auth` keeps the chinese names, and lower-casing isn't really meaningful I think? With `auth.fold(chinese).lower + year` I get `weixiao2016`. It'd be easy to add a `pinyin` filter so that you could do `auth.pinyin.lower + year`.

**Reply by retorquere** [reply]
Depending on whether you have kuroshiro turned on you will get `weixiao2016` or `giemi2016`.

**Reply by earthmoon** [reply]
> Depending on whether you have kuroshiro turned on you will get `weixiao2016` or `giemi2016`.

Now as Better Bibtex can use Chinese character or Pinyin into citation key, I needn't to use **Ideographs**  anymore, so I unclicked them. Thanks.

**Reply by earthmoon** [reply]
> My bad. Turning on `Force` will always transliterate, not remove, so I'm not sure how you got `20191` earlier. For `NRPWIZMR-refs-apse` I get `WeiXiao2016` when I turn `Force` on with `auth.lower + year`. The `lower` doesn't do anything because `auth` keeps the chinese names, and lower-casing isn't really meaningful I think? With `auth.fold(chinese).lower + year` I get `weixiao2016`. It'd be easy to add a `pinyin` filter so that you could do `auth.pinyin.lower + year`.

I am so sorry I can't reproduce the problem to show that bug. Now **the auth.fold(chinese).lower + year** will convert Chinese characters to Pinyin. So, what's the difference between **auth.fold(chinese).lower + year** and **auth.pinyin.lower**?
Thanks.

**Reply by retorquere** [reply]
No difference, `pinyin` is an alias for `fold(chinese)`. I've added it for the upcoming release, it won't work today.

**Reply by earthmoon** [reply]
> No difference, `pinyin` is an alias for `fold(chinese)`. I've added it for the upcoming release, it won't work today.

I am looking forward to use it in the future.

**Reply by retorquere** [reply]
This now works in 6.7.44

**Reply by earthmoon** [reply]
> This now works in 6.7.44

Thanks. I have updated it.
